From 7de483ebf2ad1c925d8eef5611c9ecb55ae42a1c Mon Sep 17 00:00:00 2001
From: Thomas Parrott <thomas.parrott@canonical.com>
Date: Tue, 16 Mar 2021 09:27:11 +0000
Subject: [PATCH] lxd/device/disk: Fallback to using mount device path for
 major/minor number extraction for BTRFS

If `btrfs filesystem show` doesn't work then fallback to extracting major/minor number directly from the parent block device in the mount info for cgroup limits generation.

Fixes #8564

Signed-off-by: Thomas Parrott <thomas.parrott@canonical.com>
---
 lxd/device/disk.go | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/lxd/device/disk.go b/lxd/device/disk.go
index 00fe961c501..aa46d838b92 100644
--- a/lxd/device/disk.go
+++ b/lxd/device/disk.go
@@ -1734,7 +1734,13 @@ func (d *disk) getParentBlocks(path string) ([]string, error) {
 		// Accessible btrfs filesystems
 		output, err := shared.RunCommand("btrfs", "filesystem", "show", dev[1])
 		if err != nil {
-			return nil, fmt.Errorf("Failed to query btrfs filesystem information for %q: %v", dev[1], err)
+			// Fallback to using device path to support BTRFS on block volumes (like LVM).
+			_, major, minor, errFallback := unixDeviceAttributes(dev[1])
+			if errFallback != nil {
+				return nil, errors.Wrapf(err, "Failed to query btrfs filesystem information for %q", dev[1])
+			}
+
+			devices = append(devices, fmt.Sprintf("%d:%d", major, minor))
 		}
 
 		for _, line := range strings.Split(output, "\n") {
